### PR TITLE
Find filename in upper or lower case and decode it

### PR DESCRIPTION
**Pull request by rakoo**
This PR fixes two issues:

- sometimes in the bodystructure the key is in upper case; the lib couldn't find it
- sometimes the filename is written as RFC2047-encoded words

### CODE REVIEW

**Comment by emersion** [comment]
Thanks for the patch!

> sometimes in the bodystructure the key is in upper case; the lib couldn't find it

Maybe we should lower-case keys when populating the map?

> sometimes the filename is written as RFC2047-encoded words

Ideally we'd be using `Client.options.decodeText` so that users can add support for additional charsets (by default `mime.WordDecoder` only supports UTF-8).

**Comment by rakoo** [comment]
Simon Ser ***@***.***> wrote:
> Maybe we should lower-case keys when populating the map?

Good idea, this'll make it available for other fields as well

> > sometimes the filename is written as RFC2047-encoded words
> 
> Ideally we'd be using `Client.options.decodeText` so that users can add support for additional charsets (by default `mime.WordDecoder` only supports UTF-8).

Ah indeed ! I'll use that instead


-- 
Matthieu Rakotojaona

File: imapclient/fetch.go [hunk 1]
```diff
@@ -997,23 +1004,24 @@ func readBodyFldDsp(dec *imapwire.Decoder, options *Options) (*imap.BodyStructur
...
-	}
-
 	if !dec.ExpectSpecial(')') {
 		return nil, dec.Err()
 	}
 	return &disp, nil
 }
 
-func readBodyFldParam(dec *imapwire.Decoder) (map[string]string, error) {
+func readBodyFldParam(dec *imapwire.Decoder, mimeDec func(input string) string) (map[string]string, error) {
```
Comment (emersion) [inline imapclient/fetch.go]: Can we maybe just pass the options directly instead of a function?

**Comment by rakoo** [comment]
I didn't know what would be possible, and I was kinda afraid passing the options directly would mess up the clean separation of functionalities. But if that's ok for you, I'll do it :)

-- 
Matthieu Rakotojaona

**Comment by emersion** [comment]
Yeah, I'd prefer to pass the options. It's an internal function anyways.

**Comment by rakoo** [comment]
Updated

File: imapclient/fetch.go [hunk 2]
```diff
@@ -1030,7 +1020,11 @@ func readBodyFldParam(dec *imapwire.Decoder) (map[string]string, error) {
 			if params == nil {
 				params = make(map[string]string)
 			}
-			params[k] = s
+			decoded, err := options.DecodeText(s)
+			if err != nil {
```
Comment (emersion) [inline imapclient/fetch.go]: Can we ignore the error for now, and keep the `// TODO: handle error`?

The reason is that decoding errors shouldn't be fatal, but we have no good way to bubble them up currently.

File: imapclient/fetch.go [hunk 3]
```diff
@@ -1030,7 +1020,11 @@ func readBodyFldParam(dec *imapwire.Decoder) (map[string]string, error) {
 			if params == nil {
 				params = make(map[string]string)
 			}
-			params[k] = s
+			decoded, err := options.DecodeText(s)
```
Comment (emersion) [inline imapclient/fetch.go]: CI is unhappy, seems like there is a typo here, should be `decodeText`.

Comment (rakoo) [reply imapclient/fetch.go]: Oops yes made a typo, sorry

**Comment by emersion** [comment]
Note that it's not strictly correct to decode all parameters, as encoded words are only allowed for some parameters and should be left as-is for others… But we do the same in go-message for simplicity's sake and haven't received a bug report about this yet.

**Review by emersion — APPROVED** [review]
Thanks!

**Comment by rakoo** [comment]
Thanks for your patience !

Are there really cases where some values are encoded but we don't want them decoded ? I can't imagine any, but let's see when that happens

**Pull request merged**